rustworkx-core/src/dag_algo.rs

```diff
@@ -313,6 +316,170 @@ where
     Ok(Some((path, path_weight)))
 }
 
+/// Define custom error classes for collect_bicolor_runs
+#[derive(Debug)]
+pub enum CollectBicolorError<E: Error> {
+    DAGHasCycle,
+    CallableError(E),
+}
+
+impl<E: Error> Error for CollectBicolorError<E> {}
+
+impl<E: Error> Display for CollectBicolorError<E> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            CollectBicolorError::DAGHasCycle => fmt_dag_has_cycle(f),
+            CollectBicolorError::CallableError(ref e) => fmt_callable_error(f, e),
+        }
+    }
+}
+
+fn fmt_dag_has_cycle(f: &mut Formatter<'_>) -> std::fmt::Result {
+    write!(f, "Sort encountered a cycle")
+}
+
+fn fmt_callable_error<E: Error>(f: &mut Formatter<'_>, inner: &E) -> std::fmt::Result {
+    write!(f, "The function failed with: {:?}", inner)
+}
```

Comment:
I am not 100% sure that defining these custom error classes is the best way to go, maybe it's too cumbersome? It took me a while to figure out how to convert to and from `CollectBicolorError::CallableError` and`PyErr`, and I ended up using [this function](https://github.com/Qiskit/rustworkx/pull/1186/files#diff-8d6abcabece14baab9fb4c082786ee2a676f21d9fa70162ba5222164622ff69aR610). The custom errors have also made it difficult to come up with a simple example to add to the docs, because I always need to define the type hint for the error, but this might be also lack of experience on my side.

Reply:
So I think this pattern is fine, especially if there are multiple sources of potential error. Although in the case there really is two conditions to think about, either the input graph is not a DAG (i.e. has a cycle) or the callback has an error. To simplify the function what I've typically been doing is  using `None` to represent a graph with a cycler and then using a fully generic `E` for the error type from the callback and the functions result. That lets you use `?` to pass through the callback's error directly to the caller and still handle the error case without too much extra complexity. This would basically be changing the function signature to be like:

```rust
pub fn collect_bicolor_runs<G, F, C, B, E>(
    graph: G,
    filter_fn: F,
    color_fn: C,
) -> Result<Option<Vec<Vec<G::NodeId>>>, E>
where
    F: Fn(&<G as Data>::NodeWeight) -> Result<Option<bool>, E>,
    C: Fn(&<G as Data>::EdgeWeight) -> Result<Option<usize>,E>,
    G: NodeIndexable
        + IntoNodeIdentifiers
        + IntoNeighborsDirected
        + IntoEdgesDirected
        + Visitable
        + DataMap,
    <G as GraphBase>::NodeId: Eq + Hash,
{
```
The only wrinkle in this case is because you have 2 callbacks a user needs to ensure `filter_fn` and `color_fn` return the same error type. But that should be ok in practice (and for rustworkx's python crate it will just be a `PyErr`).